optimize epg updating by updating/persisting only changes to tags...

nemphys opened the pull request:
...(instead of deleting/inserting all tags from scratch on every update)
should improve performance of epg updates due to the fact that most of the work is done in-memory.

Tested this overnight in 2 installations (both sqlite and mysql) and it worked ok in both. This is almost the first time that I check xbmc on the next day (after leaving it on overnight) and epg is not messed-up one way or another.

Please check

opdenkamp commented on xbmc/epg/Epg.cpp:
```diff
@@ -800,10 +826,18 @@ bool CEpg::PersistTags(void) const
-    at(iTagPtr)->Persist(false);
+  if (size() > 0)
+  {
+    /* For some reason, mysql fails if we use CommitInsertQueries, so we are using single updates here */
```
for what reason? what's the error?

nemphys replied:
Didn't bother to dig into that, since the function is now deprecated :-)

opdenkamp replied:
not really, since you're doing conceptually the same with your transaction

nemphys replied:
I think I remember now that CEpgInfotag->Persist() called a Delete in the database (which I removed in this commit) before the Insert/Replace. Maybe this is the reason CommitInsertQueries fails? (but then, why would it work with sqlite?)

opdenkamp replied:
yeah, the query object should only hold insert/replace/update queries. if it works/worked on sqlite3, then this was undocumented and untested behaviour

nemphys replied:
I don't see any use in the CommitInsertQueries thing (compared to proper transactions). It is quite restricting, since it is not returning unique ids (if any). Why do you avoid using proper transactions (in general)?

opdenkamp replied:
I don't have any objections against using transactions in general, I objected against _how_ you are using it

nemphys replied:
I scanned the pvr code and found only 1 call to BeginTransaction, in pvr database table creation. To me, this looks like it is being avoided ;-)

opdenkamp commented on xbmc/epg/Epg.cpp:
```diff
@@ -627,43 +652,27 @@ bool CEpg::Persist(bool bUpdateLastScanTime /* = false */)
-    epgCopy = *this;
-    m_bChanged     = false;
-    m_bTagsChanged = false;
-  }
```
why delete this? this make sure the we're not holding a lock while persisting things.

nemphys replied:
Figured that since the (long running) PersistTags has been removed, it is better to avoid the overhead of copying the tag than keeping a (very short) lock

opdenkamp replied:
you're making the locking time depend on how fast the db processes the queries.

nemphys replied:
You're right, but we might consider skipping the tags copying part for the temp epg object, since they will not be used

opdenkamp replied:
if it's not being used, we could skip this. but this isn't making much of a difference. copying a couple of tags in memory takes a couple of ms. putting that stuff in the database takes a whole lot longer. so I would suggest you focus your optimisations on that part first, and leave the epg copy trick intact for now. when you got that working 100%, you can optimise that further (possibly) be removing tricks like this. please put that in a separate commit though, since that makes it a lot easier for me to review

nemphys replied:
OK

nemphys replied:
Will update the pull request in a few minutes, including this. I just added a few more lines of code that check if the returned m_iEpgID from the database persist (of the copied epg) is the same as the real one's and if it's not update it. Otherwise, it would get lost (for new epgs I guess)

opdenkamp commented on xbmc/epg/Epg.cpp:
```diff
@@ -477,36 +481,57 @@ void CEpg::UpdateFirstAndLastDates(void)
+        CLog::Log(LOGERROR, "%s - could not open the database", __FUNCTION__);
+        return bReturn;
+      }
+      database->BeginTransaction();
```
oh please no. handle this in memory, and only send the queries that need to be sent to optimise this.

nemphys replied:
I am using a transaction in order to minimize disk access in sqlite, what's the harm?

opdenkamp replied:
you're abusing it, by first pushing all insert/update queries, then adding the delete queries and finally committing. you should clean things up internally, and then send the queries.

nemphys replied:
I don't really get what you propose here... Is the transaction thing that bothers you or something else? Do you mean that I should use single updates (ie single query transactions in sqlite) or something else?

opdenkamp replied:
I propose that you update/insert/delete all tags in memory first, then work out what changed, and then send the queries to the db. what you're doing now is starting a transaction, insert everything, delete what you don't need/want, and committing the transaction. the result is the same, but the method isn't.

also, I'd like to see all the persistence code in epgdatabase.cpp.

nemphys replied:
I am not inserting _everything_ (the previous implementation did). I am only inserting/updating what is new (or has changed) and deleting the (usually close to none) tags that need to be deleted. I'm just wrapping the whole thing in a transaction. Don't know why this is wrong. Doing it in memory would not save db queries, it would just call them from one function. I think you got it wrong, please review the code more carefuly

opdenkamp replied:
right, I'll review this again, but I thought I saw you inserting everything and cleaning up afterwards and then committing.

opdenkamp replied:
in CEpg::UpdateEntries() you open a transaction and call UpdateEntry() on each tag and passes the bStoreInDb parameter. when this is set to true, it will generate a query for each tag. next you call FixOverlappingEvents() and create delete queries for the overlapping events and commit the transaction. this sends a bunch of inserts/replaces, possibly followed by deletes for a part of the same tags.

correct me if I'm wrong.

nemphys replied:
Here is what happens with current implementation (in CEpg::UpdateEntries):
Insert _all_ tags fetched from client in memory (creates many duplicate entries)-> erase duplicate entries (FixOverlappingEvents, which actually deletes almost all of the old entries and leaves the new ones)-> persist changes in database (which also deletes all of the old entries and inserts the new ones).
In my version, the workflow is as follows:
Begin db transaction-> _Update only the tags that have changed_ and insert any new tags in memory (with changes in CEpg::UpdateEntry and CEpgInfotag::Update) and directly persist (_only the changes_)-> fix any overlapping events (the delete block there is actually never/rarely used, just left it there) and directly persist any changes to the database -> commit transaction. This should result in much less database inserts/deletes. I am wrapping the whole thing in a transaction, so as to avoid the overhead of single updates in sqlite.

opdenkamp commented:
right, please restore the epg copy trick, remove the comment in xbmc/epg/Epg.cpp:831 (it's probably because you didn't commit the queries before sending other type'd queries, like deletes, and it's not used like that anymore anyway after your changes), and I'll pull it after one final review.

nemphys committed:
optimize epg updating by updating/persisting only changes to tags (instead of deleting/inserting all tags from scratch on every update)
should improve performance of epg updates due to the fact that most of the work is done in-memory

nemphys commented:
PR updated, please review

opdenkamp commented:
thanks, merged
